---
title: "Report bulkRNAseq Analysis"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    source_code: embed
---

```{r setup, include=FALSE}
library(flexdashboard)
library(rmarkdown)
library(tinytex)
library(openxlsx)
library(readxl)
library(knitr)
library(beepr)
library(tidyverse) 
library(tximport)
library(ensembldb) 
library(EnsDb.Hsapiens.v86)
library(edgeR)
library(matrixStats)
library(cowplot)
library(svglite)
library(DT)
library(gt)
library(plotly)
library(limma)
library(gplots)
library(RColorBrewer) 
library(heatmaply)
library(tidyverse)
library(limma)
library(GSEABase)
library(Biobase)
library(GSVA)
library(gprofiler2)
library(clusterProfiler) 
library(msigdbr) 
library(enrichplot)
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```


```{r settings}
# Change title
# Change and adapt lines 343-348, 355, 417, 459
# Change line 92, data source
# Check the modules are ok on the heatmaps
# Add and change the proper study design, keep it standard

condition1<-"Sphere"
condition_control<-"Adherent"
explanation<- "Cancer Stem Cell enriched genes"
discarded_samples<-c("ERR978114","ERR978122","ERR978189","ERR978196")
nocpm_threshold<-9 #IMPORTANT PARAMETRE: minimum number of samples with CPM > 1
species<-"Homo sapiens"

```


Overview {data-icon="fa-signal"}
===================================== 

Row {data-height=400}
------------------------------------------------------------------------------

### Signal distribution: raw data

```{r raw data}
targets <- read_tsv("studydesign.txt")
all_samples<-targets$sample
samples_nodepth<-targets$sample %in% discarded_samples # remove samples without enough read depth, according to fastqc.

path <- file.path(targets$sample, "abundance.tsv") # set file paths to your mapped data
Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)
Tx <- dplyr::rename(Tx, target_id = tx_id)
Tx <- dplyr::select(Tx, "target_id", "gene_name")
Txi_gene <- tximport(paste0("kallisto/",path), 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = FALSE, #Transcripts will be collapsed to genes
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)


sampleLabels <- targets$sample
myDGEList <- DGEList(Txi_gene$counts)
log2.cpm <- cpm(myDGEList, log=TRUE)

log2.cpm.df <- as_tibble(log2.cpm, rownames = "geneID")
colnames(log2.cpm.df) <- c("geneID", sampleLabels)
log2.cpm.df.pivot <- pivot_longer(log2.cpm.df, 
                                  cols = head(targets$sample, 1):tail(targets$sample, 1),         
                                  names_to = "samples", 
                                  values_to = "expression") 
p1 <- ggplot(log2.cpm.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "Sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="unfiltered, non-normalized",
       caption=paste0("produced on ", Sys.time())) +
  scale_x_discrete(guide = guide_axis(angle = 60))+
  theme_bw()

cpm <- cpm(myDGEList)
keepers <- rowSums(cpm>1)>=nocpm_threshold 
myDGEList.filtered <- myDGEList[keepers,]

log2.cpm.filtered <- cpm(myDGEList.filtered, log=TRUE)
log2.cpm.filtered.df <- as_tibble(log2.cpm.filtered, rownames = "geneID")
colnames(log2.cpm.filtered.df) <- c("geneID", sampleLabels)
log2.cpm.filtered.df.pivot <- pivot_longer(log2.cpm.filtered.df, 
                                           cols = head(targets$sample, 1):tail(targets$sample, 1), 
                                           names_to = "samples", 
                                           values_to = "expression") 

p2 <- ggplot(log2.cpm.filtered.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "Sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="filtered, non-normalised",
       caption=paste0("produced on ", Sys.time())) +
  scale_x_discrete(guide = guide_axis(angle = 60))+
  theme_bw()

myDGEList.filtered.norm <- calcNormFactors(myDGEList.filtered, method = "TMM")
log2.cpm.filtered.norm <- cpm(myDGEList.filtered.norm, log=TRUE)
log2.cpm.filtered.norm.df <- as_tibble(log2.cpm.filtered.norm, rownames = "geneID")
colnames(log2.cpm.filtered.norm.df) <- c("geneID", sampleLabels)
log2.cpm.filtered.norm.df.pivot <- pivot_longer(log2.cpm.filtered.norm.df, 
                                                cols = head(targets$sample, 1):tail(targets$sample, 1),
                                                names_to = "samples", 
                                                values_to = "expression") 

p3 <- ggplot(log2.cpm.filtered.norm.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "Sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="filtered, TMM normalised",
       caption=paste0("produced on ", Sys.time())) +
  scale_x_discrete(guide = guide_axis(angle = 60))+
  theme_bw()

targets<-targets[!samples_nodepth,]
path <- file.path(targets$sample, "abundance.tsv") # set file paths to your mapped data
Tx <- transcripts(EnsDb.Hsapiens.v86, columns=c("tx_id", "gene_name"))
Tx <- as_tibble(Tx)
Tx <- dplyr::rename(Tx, target_id = tx_id)
Tx <- dplyr::select(Tx, "target_id", "gene_name")
Txi_gene <- tximport(paste0("kallisto/",path), 
                     type = "kallisto", 
                     tx2gene = Tx, 
                     txOut = FALSE, #Transcripts will be collapsed to genes
                     countsFromAbundance = "lengthScaledTPM",
                     ignoreTxVersion = TRUE)

sampleLabels <- targets$sample
myDGEList <- DGEList(Txi_gene$counts)
log2.cpm <- cpm(myDGEList, log=TRUE)

log2.cpm.df <- as_tibble(log2.cpm, rownames = "geneID")
colnames(log2.cpm.df) <- c("geneID", sampleLabels)
log2.cpm.df.pivot <- pivot_longer(log2.cpm.df, 
                                  cols = head(targets$sample, 1):tail(targets$sample, 1),         
                                  names_to = "samples", 
                                  values_to = "expression") 

cpm <- cpm(myDGEList)
keepers <- rowSums(cpm>1)>=nocpm_threshold 
myDGEList.filtered <- myDGEList[keepers,]

log2.cpm.filtered <- cpm(myDGEList.filtered, log=TRUE)
log2.cpm.filtered.df <- as_tibble(log2.cpm.filtered, rownames = "geneID")
colnames(log2.cpm.filtered.df) <- c("geneID", sampleLabels)
log2.cpm.filtered.df.pivot <- pivot_longer(log2.cpm.filtered.df, 
                                           cols = head(targets$sample, 1):tail(targets$sample, 1), 
                                           names_to = "samples", 
                                           values_to = "expression") 


myDGEList.filtered.norm <- calcNormFactors(myDGEList.filtered, method = "TMM")
log2.cpm.filtered.norm <- cpm(myDGEList.filtered.norm, log=TRUE)
log2.cpm.filtered.norm.df <- as_tibble(log2.cpm.filtered.norm, rownames = "geneID")
colnames(log2.cpm.filtered.norm.df) <- c("geneID", sampleLabels)
log2.cpm.filtered.norm.df.pivot <- pivot_longer(log2.cpm.filtered.norm.df, 
                                                cols = head(targets$sample, 1):tail(targets$sample, 1),
                                                names_to = "samples", 
                                                values_to = "expression") 

p4 <- ggplot(log2.cpm.filtered.norm.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "Sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="filtered, TMM normalised, fastQC passed",
       caption=paste0("produced on ", Sys.time())) +
  scale_x_discrete(guide = guide_axis(angle = 60))+
  theme_bw()
pt<-plot_grid(p1, p2, p3, p4,  labels = c('A', 'B', 'C', "D"), label_size = 12, align="hv")


p1
```

### Signal distribution: filtered and normalized data

```{r processed data}
p4
```

### PCA plot

```{r PCA}
condition <- targets$condition
condition <- factor(condition)


pca.res <- prcomp(t(log2.cpm.filtered.norm), scale.=F, retx=T)
pc.var<-pca.res$sdev^2 # sdev^2 captures these eigenvalues from the PCA result
pc.per<-round(pc.var/sum(pc.var)*100, 1) 
pca.res.df <- pca.res$x[,1:4] %>% 
  as_tibble() %>%
  add_column(sample = sampleLabels,
             group = condition)

pca.plot_condition <- ggplot(pca.res.df) +
  aes(x=PC1, y=PC2, label=sampleLabels, color = condition) +
  geom_point(size=4) +
  stat_ellipse() +
  xlab(paste0("PC1 (",pc.per[1],"%",")")) + 
  ylab(paste0("PC2 (",pc.per[2],"%",")")) +
  labs(title="PCA plot coloured by Condition",
       caption=paste0("produced on ", Sys.time())) +
  coord_fixed() +
  theme_bw()



  
pca.pivot <- pivot_longer(pca.res.df, 
                          cols = PC1:PC4, 
                          names_to = "PC", 
                          values_to = "loadings") 



ggplotly(pca.plot_condition)

```

Row {data-height=400}
------------------------------------------------------------------------------

### Filtered and normalized counts per million (CPM)

```{r filtered and normalized data}
mydata.df <- mutate(log2.cpm.filtered.norm.df,
                    adh.AVG = (ERR978118+ERR978120+ERR978121+ERR978107+ERR978112+ERR978116+ERR978117+ERR978119
)/8, 
                    sph.AVG = (ERR978193+ERR978184+ERR978185+ERR978186+ERR978187+ERR978191+ERR978194+ERR978195
)/8,
                    
                    LogFC = (sph.AVG - adh.AVG)) %>% 
  mutate_if(is.numeric, round, 2)

datatable(mydata.df[,c(1,(ncol(mydata.df)-2):ncol(mydata.df))],
          extensions = c('KeyTable'), 
          filter = 'top',
          options = list(keys = TRUE, 
                         searchHighlight = TRUE, 
                         pageLength = 10, 
                         lengthMenu = c("10", "25", "50", "100")))
```

DEGs {data-orientation=columns data-icon="fa-random"}
===================================== 

Column {data-width=500}
------------------------------------------------------------------------------

### Volcano plot

```{r volcano plot}
design <- model.matrix(~0 + condition)
colnames(design) <- levels(condition)

v.DEGList.filtered.norm <- voom(myDGEList.filtered.norm, design, plot = FALSE)
fit <- lmFit(v.DEGList.filtered.norm, design)
contrast.matrix <- makeContrasts(treatment = sph - adh,
                                 levels=design)

fits <- contrasts.fit(fit, contrast.matrix)
ebFit <- eBayes(fits)
myTopHits <- topTable(ebFit, adjust ="BH", coef=1, number=40000, sort.by="logFC")

myTopHits.df <- myTopHits %>%
  as_tibble(rownames = "geneID")


vplot <- ggplot(myTopHits.df) +
  aes(y=-log10(adj.P.Val), x=logFC, text = paste("Symbol:", geneID)) +
  geom_point(size=2) +
  geom_hline(yintercept = -log10(0.01), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 2, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -2, linetype="longdash", colour="#2C467A", size=1) +
  #annotate("rect", xmin = 1, xmax = 12, ymin = -log10(0.01), ymax = 7.5, alpha=.2, fill="#BE684D") +
  #annotate("rect", xmin = -1, xmax = -12, ymin = -log10(0.01), ymax = 7.5, alpha=.2, fill="#2C467A") +
  labs(title="Volcano plot",
       subtitle = explanation,
       caption=paste0("produced on ", Sys.time())) +
  scale_x_continuous(n.breaks = 8)+
  theme_bw()

# Only in html ggplotly(vplot)
ggplotly(vplot)

```


Column {.tabset data-width=500}
------------------------------------------------------------------------------

### DEG table

```{r DEG table}
results <- decideTests(ebFit, method="global", adjust.method="BH", p.value=0.01, lfc=2)
colnames(v.DEGList.filtered.norm$E) <- sampleLabels
diffGenes <- v.DEGList.filtered.norm$E[results[,1] !=0,]
diffGenes.df <- as_tibble(diffGenes, rownames = "geneID")
datatable(diffGenes.df, 
          extensions = c('KeyTable'),
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 10, lengthMenu = c("10", "25", "50", "100"))) %>%
  formatRound(columns=c(2:ncol(diffGenes.df)), digits=2)
```

### Heatmap

```{r heatmap}
myheatcolors <- rev(brewer.pal(name="RdBu", n=11))
clustRows <- hclust(as.dist(1-cor(t(diffGenes), method="pearson")), method="complete") 

clustColumns <- hclust(as.dist(1-cor(diffGenes, method="spearman")), method="complete")
module.assign <- cutree(clustRows, k=2) #Important parametre
module.color <- rainbow(length(unique(module.assign)), start=0.1, end=0.9) 
module.color <- module.color[as.vector(module.assign)] 

heatmaply(diffGenes, 
          colors = myheatcolors,
          Rowv=as.dendrogram(clustRows),
          RowSideColors=module.color,
          showticklabels=c(TRUE,FALSE),
          scale='row',
          main ="Heatmap across all samples")

```

Functional enrichment {data-icon="fa-signal"}
===================================== 

Column {.tabset data-width=500}
------------------------------------------------------------------------------

### GO - enriched in `r condition1`

```{r gostplot1}
modulePick <- 1  #Be careful with which is which
myModule_up <- diffGenes[names(module.assign[module.assign %in% modulePick]),] 
hrsub_up <- hclust(as.dist(1-cor(t(myModule_up), method="pearson")), method="complete") 

gost.res_up <- gost(rownames(myModule_up), organism = "hsapiens", correction_method = "fdr")
gostplot(gost.res_up, interactive = T, capped = T) #set interactive=FALSE to get plot for publications
```

### GO - enriched in `r condition_control`

```{r gostplot2}
modulePick <- 2 
myModule_down <- diffGenes[names(module.assign[module.assign %in% modulePick]),] 
hrsub_down <- hclust(as.dist(1-cor(t(myModule_down), method="pearson")), method="complete") 

gost.res_down <- gost(rownames(myModule_down), organism = "hsapiens", correction_method = "fdr")
gostplot(gost.res_down, interactive = T, capped = T) #set interactive=FALSE to get plot for publications
```

Column {data-width=500}
------------------------------------------------------------------------------

### GSEA C2 Collection

```{r GSEA}
hs_gsea_c2 <- msigdbr(species = "Homo sapiens",
                      category = "C2") %>% 
  dplyr::select(gs_name, gene_symbol)

mydata.df.sub <- dplyr::select(mydata.df, geneID, LogFC)
mydata.gsea <- mydata.df.sub$LogFC
names(mydata.gsea) <- as.character(mydata.df.sub$geneID)
mydata.gsea <- sort(mydata.gsea, decreasing = TRUE)

# run GSEA using the 'GSEA' function from clusterProfiler
myGSEA.res <- GSEA(mydata.gsea, TERM2GENE=hs_gsea_c2, verbose=FALSE)
myGSEA.df <- as_tibble(myGSEA.res@result)

datatable(myGSEA.df, 
          extensions = c('KeyTable'), 
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 10, lengthMenu = c("10", "25", "50", "100")))
```

### GSEA Hallmark Collection

```{r GSEA}
hs_gsea_H <- msigdbr(species = "Homo sapiens",
                      category = "H") %>% 
  dplyr::select(gs_name, gene_symbol)


# run GSEA using the 'GSEA' function from clusterProfiler
myGSEA.res_H <- GSEA(mydata.gsea, TERM2GENE=hs_gsea_H, verbose=FALSE)
myGSEA.df_H <- as_tibble(myGSEA.res@result)

datatable(myGSEA.df_H, 
          extensions = c('KeyTable'), 
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 10, lengthMenu = c("10", "25", "50", "100")))
```


About {data-orientation=rows data-icon="fa-comment-alt"}
===================================== 

Row {data-height=1000}
------------------------------------------------------------------------------

Use this page to describe the results in your own words.  Some things to think about:

* What are the key takeaways from the analysis?
* What types of analyses would you want to do next?
* Based on your analysis, what wet-lab experiments would you pursue?
* How could you expand on or otherwise enhance this data dashboard?

